torch/hub.py

```diff
@@ -59,112 +98,285 @@ def _download_url_to_file(url, filename):
+
+
+def _check_module_exists(name):
+    if sys.version_info >= (3, 4):
+        import importlib.util
+        return importlib.util.find_spec(name) is not None
+    elif sys.version_info >= (3, 3):
+        # Special case for python3.3
+        import importlib.find_loader
+        return importlib.find_loader(name) is not None
+    else:
+        # NB: imp doesn't handle hierarchical module names (names contains dots).
+        try:
+            import imp
+            imp.find_module(name)
+        except Exception:
+            return False
+        return True
+
 
-    dependencies = _load_attr_from_module('hubconf', 'dependencies')
+def _check_dependencies(m):
+    dependencies = _load_attr_from_module(m, VAR_DEPENDENCY)
 
     if dependencies is not None:
         missing_deps = [pkg for pkg in dependencies if not _check_module_exists(pkg)]
         if len(missing_deps):
             raise RuntimeError('Missing dependencies: {}'.format(', '.join(missing_deps)))
 
-    func = _load_attr_from_module('hubconf', model)
-    if func is None:
+
+def _load_entry_from_hubconf(m, model):
+    if not isinstance(model, str):
+        raise ValueError('Invalid input: model should be a string of function name')
+
+    _check_dependencies(m)
+
+    func = _load_attr_from_module(m, model)
+
+    if func is None or not callable(func):
         raise RuntimeError('Cannot find callable {} in hubconf'.format(model))
 
-    # Check if func is callable
-    if not callable(func):
-        raise RuntimeError('{} is not callable'.format(func))
+    return func
+
+
+def set_dir(d):
+    r"""
+    Optionally set hub_dir to a local dir to save downloaded models & weights.
+
+    If this argument is not set, env variable `TORCH_HUB_DIR` will be searched first,
+    `~/.torch/hub` will be created and used as fallback.
+
+    Args:
+        d: path to a local folder to save downloaded models & weights.
+    """
+    global hub_dir
+    hub_dir = d
+
+
+def list(github, force_reload=False):
+    r"""
+    List all entrypoints available in `github` hubconf.
+    Args:
+        github: Required, a string with format "repo_owner/repo_name[:tag_name]" with an optional
+            tag/branch. The default branch is `master` if not specified.
+            Example: 'pytorch/vision[:hub]'
+        force_reload: Optional, whether to discard the existing cache and force a fresh download.
+            Default is `False`.
+    Returns:
+        entrypoints: a list of available entrypoint names
+
+    Example:
+        >>> entrypoints = torch.hub.list('pytorch/vision', force_reload=True)
+    """
+
+    hub_module = _load_hubconf_module(github, force_reload)
+
+    # We take functions starts with '_' as internal helper functions
+    entrypoints = [f for f in dir(hub_module) if callable(getattr(hub_module, f)) and not f.startswith('_')]
+
+    return entrypoints
+
 
-    # Call the function
-    return func(*args, **kwargs)
+def show(github, model, force_reload=False):
```

Comment:
i think it's better to return `torch.hub.help(...)` instead of `torch.hub.show(...)` and instead of printing `__doc__`, just returning it.
The usage is not that different in the end, and the user is in control of the print.
What do you think? (I am a bit split about this)

Reply:
lol I'm also super split on this and I changed it back and forth. I definitely agree that returning it a string would make it more manageable for users. And then I can't think of another way user can use it other than printing it out :P That's why I changed. Yea I can make it return the string instead of printing and let's see if we get feedback on it. :D